### ISSUE TITLE
GRPC error: failed to mount secrets store objects for pod <Pod_Name>, err: rpc error: code = Canceled desc = context canceled

### ISSUE THREAD

**Issue by ArronaxKP** [Azure/secrets-store-csi-driver-provider-azure#471]
**Have you**

- [x] Read [Troubleshooting Guide](https://azure.github.io/secrets-store-csi-driver-provider-azure/troubleshooting/)
- [x] Searched on [GitHub issues](https://github.com/Azure/secrets-store-csi-driver-provider-azure/issues) and [Discussions](https://github.com/Azure/secrets-store-csi-driver-provider-azure/discussions)

**What steps did you take and what happened:**
So we are using the CSI driver and it seems to work as expected for extended periods of time. We are running the helm chart version 0.0.18.

Then randomly we start having pods fail to start (a re-deployment of the deployment changing the image version is normally the root cause).
Then the pod will keep restarting every 2 minutes for 10 minutes then it seems to come back to life.

The CSI Driver log is showing this:
```
1 utils.go:79] GRPC error: failed to mount secrets store objects for pod <namespace>/<pod-name>, err: rpc error: code = Canceled desc = context canceled
E0412 15:02:43.926932       1 utils.go:79] GRPC error: failed to mount secrets store objects for pod <namespace>/<pod-name>, err: rpc error: code = Canceled desc = context canceled
E0412 15:04:45.022779       1 utils.go:79] GRPC error: failed to mount secrets store objects for pod <namespace>/<pod-name>, err: rpc error: code = Canceled desc = context canceled
E0412 15:06:47.110757       1 utils.go:79] GRPC error: failed to mount secrets store objects for pod <namespace>/<pod-name>, err: rpc error: code = Canceled desc = context canceled
E0412 15:08:51.182836       1 utils.go:79] GRPC error: failed to mount secrets store objects for pod <namespace>/<pod-name>, err: rpc error: code = Canceled desc = context canceled
E0412 15:10:59.270827       1 utils.go:79] GRPC error: failed to mount secrets store objects for pod <namespace>/<pod-name>, err: rpc error: code = Canceled desc = context canceled
```

The Provider Log is:   
```
E0412 15:00:43.321086       1 server.go:54] "failed to process mount request" err="failed to get objectType:secret, objectName:keyvault-secret-id, objectVersion:: keyvault.BaseClient#GetSecret: Failure sending request: StatusCode=0 -- Original Error: context canceled"
E0412 15:00:43.321126       1 grpc.go:29] "GRPC error" err="failed to mount objects, error: failed to get objectType:secret, objectName:keyvault-secret-id, objectVersion:: keyvault.BaseClient#GetSecret: Failure sending request: StatusCode=0 -- Original Error: context canceled"
I0412 15:00:43.908271       1 provider.go:247] "unmarshaled key vault objects" keyVaultObjects=[{VALID_JSON_PAYLOAD}] count=16 pod="<namespace>/<pod-name>"
I0412 15:00:43.908301       1 provider.go:258] "fetching object from key vault" objectName="keyvault-secret-id" objectType="secret" keyvault="keyvault-name" pod="<namespace>/<pod-name>"
I0412 15:00:43.908554       1 auth.go:146] "using service principal to retrieve access token" clientID="########## REDACTED ##########" secret="########## REDACTED ##########" pod="<namespace>/<pod-name>"
E0412 15:02:43.900860       1 server.go:54] "failed to process mount request" err="failed to get objectType:secret, objectName:keyvault-secret-id, objectVersion:: keyvault.BaseClient#GetSecret: Failure sending request: StatusCode=0 -- Original Error: context canceled"
E0412 15:02:43.900914       1 grpc.go:29] "GRPC error" err="failed to mount objects, error: failed to get objectType:secret, objectName:keyvault-secret-id, objectVersion:: keyvault.BaseClient#GetSecret: Failure sending request: StatusCode=0 -- Original Error: context canceled"
I0412 15:02:45.006541       1 provider.go:247] "unmarshaled key vault objects" keyVaultObjects=[{VALID_JSON_PAYLOAD}] count=16 pod="<namespace>/<pod-name>"
I0412 15:02:45.006577       1 provider.go:258] "fetching object from key vault" objectName="keyvault-secret-id" objectType="secret" keyvault="keyvault-name" pod="<namespace>/<pod-name>"
I0412 15:02:45.006825       1 auth.go:146] "using service principal to retrieve access token" clientID="########## REDACTED ##########" secret="########## REDACTED ##########" pod="<namespace>/<pod-name>"
```

**What did you expect to happen:**
Unsure if this is something in the CSI driver or we are actually seeing a symptom from something else and the CSI driver happens to be the only thing logging (as I am not seeing an error caused by the CSI driver)

**Anything else you would like to add:**
Nothing additional, we are using the stock resource allocation for the CSI driver pods and the provider pod but our cluster is host 10s of applications per node so should we also be increasing the resource allocation

**Which access mode did you use to access the Azure Key Vault instance:**
Authentication is currently done using a Service Principal

**Environment:**
Using helm chart version: 0.0.18
AKS Version: 1.17.13
Linux Node pool: Standard_E8s_v3

**Comment by aramase** [comment]
@ArronaxKP This issue looks very similar to https://github.com/Azure/secrets-store-csi-driver-provider-azure/issues/292

1. Do you have a firewall rule blocking egress from the host?
  - The provider pods request to get a token and access keyvault is being blocked based on the logs posted.
2. Could you check if there is connectivity to AAD and keyvault from the node?

**Comment by ArronaxKP** [comment]
Yes our symptoms look the same as #292. I will check our side but network issues is entirely possible. Will look through our side to see if we can identify the root cause.
As #292 has the same error and was host network related I would our issue is also network connectivity related. Will update this once I identify what component was failing to connect.

**Comment by aramase** [comment]
@ArronaxKP Any updates on this?

**Comment by ArronaxKP** [comment]
So we haven't been able to pin-point the root cause. Its seems mostly stable. As we have seen some intermittent network issues with where this is deployed I am going to assume that is the issue and close this issue.

**Comment by ohhelloxoxo** [comment]
I see this was closed, but I did want to throw in my two cents. I'm setting up the CSI driver/Azure provider for the first time on my cluster today, and I've gotten everything setup except for running into this same issue.

For what it's worth, I don't seem to have any firewall or connectivty issues: the timeout for the SP trying to get a token is consistent, and on that same pod I can use wget and get an AAD token instantly with no issue using the same clientId/password/tenantId.
